File: js/src/chunking.ts
/**
 * Punctuation normalization + sentence chunking — port of zerotts/chunking.py.
 *
 * The model is trained on utterances, not paragraphs, and generation is bounded
 * by `maxFrames` (1500 frames = 120 s). Feeding it a whole article as one
 * utterance does not error — it just stops mid-sentence when the frame budget
 * runs out. Long text has to be split.
 *
 * Splitting falls back through a hierarchy so no segment exceeds the budget,
 * however the input is (or is not) punctuated:
 *   1. sentence-ending punctuation (. ! ? …)
 *   2. commas, if a "sentence" is still too long
 *   3. word boundaries, if a comma-separated piece is still too long
 *   4. raw character slicing, for a single unbroken run
 *
 * Keep this in step with the Python version: the two are expected to segment
 * identically, and a divergence shows up as the browser and the package
 * producing different audio for the same input.
 */

const SENTENCE_END = /(?<=[.!?…])\s+/;
const COMMA = /(?<=,)\s+/;

/** Rough speaking rate, used to size segments by estimated duration. */
const CHARS_PER_SEC = 15.0;

const EXISTING_PUNCT = new Set([...'.!?…,;:—–-"\')]}']);

/**
 * Rewrite written-only punctuation into breaks the model can voice.
 *
 *   ';' -> ','       the vocab has the token, but the training corpus is
 *                    transcribed speech where it barely occurs, so the model has
 *                    no reliable prosody for it.
 *   newline -> '. '  unless the line already ends in punctuation. The tokenizer
 *                    collapses all whitespace to a single space, so an
 *                    un-rewritten line break vanishes entirely and two unrelated
 *                    lines run together in one breath.
 *
 * Semicolons are converted first, so a line ending in one already counts as
 * punctuated when the newline rule runs and does not also collect a '.'.
 */
export function normalizePunctuation(text: string): string {
  if (!text) return text;
  let out = text.replace(/;/g, ',');
  out = out.replace(/(.?)[ \t]*(?:\r?\n[ \t]*)+/g, (_m, prev: string) => {
    if (!prev) return '';                       // leading newline: nothing to punctuate
    if (EXISTING_PUNCT.has(prev)) return `${prev} `;
    return `${prev}. `;
  });
  return out.trim();
}

function splitSentences(text: string): string[] {
  const trimmed = text.trim();
  if (!trimmed) return [];
  const sentences: string[] = [];
  for (const para of trimmed.split(/\n\s*\n/)) {
    const p = para.trim();
    if (!p) continue;
    for (const sent of p.split(SENTENCE_END)) {
      const s = sent.trim();
      if (s) sentences.push(s);
    }
  }
  return sentences;
}

/** Greedily join pieces (each already <= maxChars) as close to maxChars as possible. */
function packPieces(pieces: string[], maxChars: number): string[] {
  const chunks: string[] = [];
  let current: string[] = [];
  let currentLen = 0;
  for (const piece of pieces) {
    let pieceLen = piece.length + (current.length ? 1 : 0);
    if (current.length && currentLen + pieceLen > maxChars) {
      chunks.push(current.join(' '));
      current = [];
      currentLen = 0;
      pieceLen = piece.length;
    }
    current.push(piece);
    currentLen += pieceLen;
  }
  if (current.length) chunks.push(current.join(' '));
  return chunks;
}

function splitByChars(text: string, maxChars: number): string[] {
  const out: string[] = [];
  for (let i = 0; i < text.length; i += maxChars) out.push(text.slice(i, i + maxChars));
  return out;
}

function splitByWords(text: string, maxChars: number): string[] {
  const packed = packPieces(text.split(/\s+/).filter(Boolean), maxChars);
  const out: string[] = [];
  for (const piece of packed) {
    if (piece.length <= maxChars) out.push(piece);
    else out.push(...splitByChars(piece, maxChars));
  }
  return out;
}

/** Break text down until every piece is <= maxChars, least disruptive split first. */
function atomize(text: string, maxChars: number): string[] {
  const t = text.trim();
  if (!t) return [];
  if (t.length <= maxChars) return [t];

  const commaParts = t.split(COMMA).map((p) => p.trim()).filter(Boolean);
  if (commaParts.length > 1) {
    const out: string[] = [];
    for (const part of commaParts) out.push(...atomize(part, maxChars));
    return out;
  }
  if (/\s/.test(t)) return splitByWords(t, maxChars);
  return splitByChars(t, maxChars);
}

export function chunkText(text: string, maxChunkSec = 15.0): string[] {
  const maxChars = Math.max(1, Math.floor(maxChunkSec * CHARS_PER_SEC));
  const atoms: string[] = [];
  for (const sentence of splitSentences(text)) atoms.push(...atomize(sentence, maxChars));
  return packPieces(atoms, maxChars);
}

// ── per-segment punctuation cleanup ─────────────────────────────────────────
// Chunking can cut a sentence mid-flow, leaving a segment that ends on a comma
// or nothing, and/or a stray terminal mark stranded in the middle of what is now
// a standalone utterance. Interior punctuation reads as a hard stop where only a
// pause belongs, so mid-segment marks are downgraded to a comma-pause and the
// segment's own end gets a real terminator.

const TRAILING = /[^\p{L}\p{N}_]+$/u;
const MID_PUNCT = /[^\p{L}\p{N}_\s\-/.,:?@!"'%]/gu;
const REPEAT_COMMA = /\s*(?:,\s*)+/g;
const END_PUNCT = new Set(['.', '!', '?']);

export function cleanSegmentPunctuation(text: string): string {
  const t = text.trim();
  if (!t) return t;

  const match = t.match(TRAILING);
  const core0 = match ? t.slice(0, match.index) : t;
  const trailing = match ? (match[0] ?? '').trimEnd() : '';

  const endPunct = trailing && END_PUNCT.has(trailing[trailing.length - 1])
    ? trailing[trailing.length - 1]
    : '.';

  let core = core0.replace(MID_PUNCT, ',');
  core = core.replace(REPEAT_COMMA, ', ').replace(/^[\s,]+|[\s,]+$/g, '');
  if (!core) return '';
  return core + endPunct;
}

/** The exact segments the synthesizer will speak, chunked and cleaned. */
export function textSegments(text: string, maxChunkSec = 15.0): string[] {
  return chunkText(normalizePunctuation(text), maxChunkSec)
    .map(cleanSegmentPunctuation)
    .filter(Boolean);
}

/** Parse a "### name"-delimited sample-text file into {name: text}. */
export function parseSamples(raw: string): Record<string, string> {
  const samples: Record<string, string> = {};
  let name: string | null = null;
  let lines: string[] = [];
  const flush = () => {
    if (name !== null) samples[name] = lines.join('\n').trim();
  };
  for (const line of raw.split(/\r?\n/)) {
    if (line.startsWith('### ')) {
      flush();
      name = line.slice(4).trim();
      lines = [];
    } else if (line.startsWith('#')) {
      continue;
    } else {
      lines.push(line);
    }
  }
  flush();
  return samples;
}

File: src/zerotts/chunking.py
"""Punctuation normalization + sentence segmentation/chunking for long-form
text-to-speech input.

No sentence-boundary-detection library is a project dependency (no nltk/
spacy/pysbd), so this uses a regex-based splitter, then greedily packs
sentences into chunks bounded by an estimated speaking duration.

Splitting falls back through a hierarchy so no single chunk ever exceeds
the character budget, however the user punctuates (or doesn't):
    1. sentence-ending punctuation (. ! ? …)
    2. commas, if a "sentence" is still too long
    3. word boundaries, if a comma-separated piece is still too long
    4. raw character slicing, as a last resort (e.g. one giant run-on word)
"""

from __future__ import annotations

import re

# Split after ., !, ?, or … followed by whitespace (or end of string), but
# not on a run of digits/abbreviation-like single letters (e.g. "3.5" or
# "Mr.") immediately before the dot — kept intentionally simple/heuristic.
_SENTENCE_END_RE = re.compile(r"(?<=[.!?…])\s+")

# Split after a comma followed by whitespace, keeping the comma with the
# preceding text (used only when a whole "sentence" is over budget).
_COMMA_RE = re.compile(r"(?<=,)\s+")

# Rough average speaking rate used to size chunks by estimated duration
# rather than raw character count (TTS speaking rate varies by language,
# so this is a coarse heuristic, not a precise duration predictor).
_CHARS_PER_SEC = 15.0

# ── punctuation normalization ────────────────────────────────────────────────
# Written text carries breaks the model was never trained to voice. The BPE
# vocab has a token for ';', but the training corpus is transcribed speech,
# where a semicolon is rare — the model has no reliable prosody for it, and at
# inference it tends to read as a hard stop or as nothing at all. A comma is
# the break it actually stands for out loud, so it is rewritten to one. ':' is
# left as-is (kept, not folded to a comma). Newlines are worse: the BPE
# normalizer collapses all whitespace to a single space
# (zerotts.tokenizer.normalize_text), so a line break vanishes entirely
# and two unrelated lines run together as one breath. Turning it into a
# sentence stop preserves the pause the layout meant — unless the line already
# ends in punctuation, which already carries it.

_PAUSE_PUNCT_RE = re.compile(r"[;]")

# A newline (plus any surrounding horizontal whitespace, and any run of blank
# lines) with what immediately precedes it captured, so the replacement can look
# at whether that character is already punctuation.
_NEWLINE_RE = re.compile(r"(.?)[ \t]*(?:\r?\n[ \t]*)+")

# Characters that already terminate or pause a clause — a line ending in one of
# these needs no added '.'.
_EXISTING_PUNCT = set(".!?…,;:—–-\"')]}")


def normalize_punctuation(text: str) -> str:
    """Rewrite written-only punctuation into the breaks the model can voice.

        ';' -> ','
        newline -> '. ' when the line does not already end in punctuation,
                   otherwise just a space.

    Semicolons are converted FIRST, so a line ending in one counts as
    already-punctuated when the newline rule runs (it ends in ',' by then)
    and does not also collect a '.'. ':' is untouched by this step, but is
    already in _EXISTING_PUNCT so a line ending in one is also treated as
    already-punctuated.
    """
    if not text:
        return text

    text = _PAUSE_PUNCT_RE.sub(",", text)

    def _repl(m: re.Match) -> str:
        prev = m.group(1)
        if not prev:  # leading newline(s) — nothing before them to punctuate
            return ""
        if prev in _EXISTING_PUNCT:
            return f"{prev} "
        return f"{prev}. "

    text = _NEWLINE_RE.sub(_repl, text)
    return text.strip()


def split_sentences(text: str) -> list[str]:
    text = text.strip()
    if not text:
        return []
    # Split paragraphs first so blank lines always force a break, then
    # sentence-split within each paragraph.
    sentences: list[str] = []
    for para in re.split(r"\n\s*\n", text):
        para = para.strip()
        if not para:
            continue
        for sent in _SENTENCE_END_RE.split(para):
            sent = sent.strip()
            if sent:
                sentences.append(sent)
    return sentences


def _pack_pieces(pieces: list[str], max_chars: int) -> list[str]:
    """Greedily join `pieces` (each already <= max_chars) with a single
    space, filling each output chunk as close to max_chars as possible
    without exceeding it."""
    chunks: list[str] = []
    current: list[str] = []
    current_len = 0
    for piece in pieces:
        piece_len = len(piece) + (1 if current else 0)
        if current and current_len + piece_len > max_chars:
            chunks.append(" ".join(current))
            current, current_len = [], 0
            piece_len = len(piece)
        current.append(piece)
        current_len += piece_len
    if current:
        chunks.append(" ".join(current))
    return chunks


def _split_by_chars(text: str, max_chars: int) -> list[str]:
    return [text[i:i + max_chars] for i in range(0, len(text), max_chars)]


def _split_by_words(text: str, max_chars: int) -> list[str]:
    words = text.split()
    pieces = _pack_pieces(words, max_chars)
    # A single word longer than max_chars (e.g. a spammed run of characters
    # with no spaces) still needs a hard character split.
    out: list[str] = []
    for piece in pieces:
        if len(piece) <= max_chars:
            out.append(piece)
        else:
            out.extend(_split_by_chars(piece, max_chars))
    return out


def _atomize(text: str, max_chars: int) -> list[str]:
    """Break `text` down until every returned piece is <= max_chars,
    preferring the least disruptive split available (comma > word >
    character)."""
    text = text.strip()
    if not text:
        return []
    if len(text) <= max_chars:
        return [text]

    comma_parts = [p.strip() for p in _COMMA_RE.split(text) if p.strip()]
    if len(comma_parts) > 1:
        out: list[str] = []
        for part in comma_parts:
            out.extend(_atomize(part, max_chars))
        return out

    if " " in text.strip():
        return _split_by_words(text, max_chars)

    return _split_by_chars(text, max_chars)


def chunk_text(text: str, max_chunk_sec: float = 15.0) -> list[str]:
    """Split `text` into speakable chunks whose estimated duration stays
    near `max_chunk_sec`, preferring to break on sentence-ending
    punctuation, then commas, then words, then raw characters as a last
    resort — so a single chunk never balloons past the budget no matter how
    the input is (or isn't) punctuated.
    """
    max_chars = max(1, int(max_chunk_sec * _CHARS_PER_SEC))
    sentences = split_sentences(text)

    atoms: list[str] = []
    for sent in sentences:
        atoms.extend(_atomize(sent, max_chars))

    return _pack_pieces(atoms, max_chars)


# ── per-segment punctuation cleanup ──────────────────────────────────────────
# Chunking (above) can cut a sentence mid-flow, leaving a segment that ends on
# a comma or nothing at all, and/or a stray '.'/'!'/'?'/other symbol stranded
# in the middle of what is now a shorter, standalone utterance. Interior
# punctuation reads as a hard stop (or gets silently dropped by the tokenizer)
# where only a pause belongs, so every mid-segment punctuation mark other than
# a hyphen — word-joining ('twenty-one') or otherwise —, a slash, or a dot is
# downgraded to a comma-pause. The segment's own end is where a real stop belongs: whatever
# terminal mark it already had is kept if it's one of '.', '!', '?' (that's
# real, meaningful prosody), and anything else trailing there — comma, dash,
# stray symbol, nothing — is normalized to a single '.'.

# Trailing run of whitespace/punctuation, i.e. everything after the last
# alphanumeric character in the segment.
_TRAILING_RE = re.compile(r"[^\w]+$", re.UNICODE)

# Any mid-segment character that is not a word char, whitespace, hyphen,
# slash, dot, colon, question mark, exclamation mark, quote (single or
# double), percent sign, or comma itself gets folded into a comma-pause.
_MID_PUNCT_RE = re.compile(r"[^\w\s\-/.,:?@!\"'%]", re.UNICODE)

# Collapse runs of commas (possibly separated by spaces) into a single
# ', ' — also fixes the missing space left when a punctuation mark abutting
# a word (e.g. the '"' in 'Hello"world') is folded straight into a comma.
_REPEAT_COMMA_RE = re.compile(r"\s*(?:,\s*)+")

_END_PUNCT = {".", "!", "?"}


def clean_segment_punctuation(text: str) -> str:
    """Normalize a single chunked segment's punctuation for TTS:

        1. every mid-segment punctuation mark other than
           '-', '/', '.', ':', '?', '!', '"', "'", '%' becomes a ',' pause
        2. terminal mark: kept as-is if it's '.', '!', or '?', otherwise
           whatever trails (comma, dash, stray symbol, nothing) -> '.'
    """
    text = text.strip()
    if not text:
        return text

    m = _TRAILING_RE.search(text)
    core = text[:m.start()] if m else text
    trailing = text[m.start():].rstrip() if m else ""

    end_punct = trailing[-1] if trailing and trailing[-1] in _END_PUNCT else "."

    core = _MID_PUNCT_RE.sub(",", core)
    core = _REPEAT_COMMA_RE.sub(", ", core).strip(" ,")

    if not core:
        return ""
    return core + end_punct


# ── sample text file loading ─────────────────────────────────────────────────
# Used by the Gradio demo's "Sample texts" picker.

def load_text_samples(path) -> dict[str, str]:
    """Parse a "### name" delimited sample-text file (see webui/test_samples.txt's
    header for the format) into {name: text}. Missing file -> {}."""
    from pathlib import Path

    path = Path(path)
    if not path.is_file():
        return {}

    samples: dict[str, str] = {}
    name = None
    lines: list[str] = []

    def flush():
        if name is not None:
            samples[name] = "\n".join(lines).strip()

    for raw_line in path.read_text(encoding="utf-8").splitlines():
        if raw_line.startswith("### "):
            flush()
            name = raw_line[4:].strip()
            lines = []
        elif raw_line.startswith("#"):
            continue
        else:
            lines.append(raw_line)
    flush()
    return samples
